src/thread.ts

```diff
@@ -135,7 +135,26 @@ export default class Thread {
         } else if (type === 'boolean') {
             this.cmodule.lua_pushboolean(this.address, target)
         } else if (type === 'object') {
-            if (target instanceof Thread) {
+            if (target instanceof Promise) {
+                this.pushValue({
+                    next: (_: unknown, ...args: Parameters<typeof target.then>) => target.then(...args),
+                    await: decorateFunction(
+                        (thread: Thread) => {
+                            // eslint-disable-next-line
+                            target.then((result: any) => {
+                                thread.pushValue(result)
+                                this.cmodule.lua_resume(thread.address, this.address, 1)
```

Comment:
Perhaps missing a catch or finally here? If the target throws I don't think lua_resume will be called?

Reply:
The fourth argument to lua_resume is the number of returned args. These need to be popped before resuming, I'd recommend a check and pop here for safety

Reply:
> Perhaps missing a catch or finally here? If the target throws I don't think lua_resume will be called?

I added a second return for await to handle errors:
```
local res, error = promise:await()
```

Would be nice to have the default lua error management, but I realized pcall doesn't work well with yields :/

Reply:
It can if you catch the error and then raise it as a lua_error

Reply:
I tried exactly that, but no success :/

Reply:
Fyi once outside of resume you can't call lua_error because you're not within the Lua VM. You can do it from the continuance function though

Reply:
@timstableford thx for the tip! I changed it and now we can use pcall